### ISSUE TITLE
[RFC]  vLLM-Kunlun Roadmap Q1 2026

### ISSUE THREAD

**Issue by xyDong0223** [baidu/vLLM-Kunlun#156]
This is a living document! For each item here, we intend to link the RFC

*In Q4 2025, we covered most mainstream models (Qwen3-next, DeepSeek v3.2, GLM4.7, Qwen3VL, InternVL, etc.) and enriched the quantization ecosystem, supporting INT8 W8A8 and W4A16 deployment for both dense and MoE models.*

<img width="1736" height="692" alt="Image" src="https://github.com/user-attachments/assets/ec982cce-6877-4b79-94f4-62349f132103" />

![](https://rte.weiyun.baidu.com/wiki/attach/image/api/imageDownloadAddress?attachId=70731850c35240dc92ee27eafd2078e4&docGuid=vaFMwY5ShRz_Ib)
*We expect to further expand the range of supported models and sampling parameters in Q1 2026, and to support xgrammar calls, enabling native function call capabilities.*

### vLLM version update
- [x] update v0.11.0 to v0.15.1  https://github.com/baidu/vLLM-Kunlun/issues/236
- [ ] update model runner to v2

### Model support
- [x] GLM5  https://github.com/baidu/vLLM-Kunlun/pull/194
- [x] Mimimax 2.1 https://github.com/baidu/vLLM-Kunlun/pull/275
- [ ] Kimi2
- [x] Qwen3.5 https://github.com/baidu/vLLM-Kunlun/pull/241

### Sampling parameters support
- [ ] min_tokens

### Feature
- [ ] int8 w4a8
- [x] Function call  https://github.com/baidu/vLLM-Kunlun/pull/281

### OOT
- [x] Make Cluster Ops OOT  https://github.com/baidu/vLLM-Kunlun/pull/234 
- [x] Make Moe OOT https://github.com/baidu/vLLM-Kunlun/pull/209

**Comment by xueliangyang-oeuler** [comment]
As for updating v0.11.0 to v0.14.1, are there any specific features to be supported in vllm 0.14.1, such as overlap-schedule, speculative decoding? Thank you.

**Comment by GrootLiu** [comment]
I observed that vllm-kunlun created a new folder to store all the models, and it requires using these to replace the native vllm model files. Is it possible to use the existing model files from the vllm project instead of creating new ones in the current project? Otherwise, the maintenance overhead will be very high.

**Comment by xyDong0223** [comment]
> I observed that vllm-kunlun created a new folder to store all the models, and it requires using these to replace the native vllm model files. Is it possible to use the existing model files from the vllm project instead of creating new ones in the current project? Otherwise, the maintenance overhead will be very high.

Thanks for the suggestion! We have already removed the Llama and Qwen2 series model files, and we plan to remove more to reduce overhead. If you have any specific recommendations on which models we should prioritize for removal, please let us know!

**Comment by xyDong0223** [comment]
# vLLM-Kunlun v0.15.1 Framework Upgrade Summary

## 1. Background

[vLLM Kunlun (vllm-kunlun)](https://github.com/baidu/vLLM-Kunlun) is a community-maintained hardware plugin that enables vLLM to run seamlessly on Baidu's Kunlun XPU. The upstream vLLM v0.15.1 release introduced several breaking architectural changes:

- **V1-Only Architecture**: The V0/V1 dual execution path was removed; the framework is now V1-only
- **Module Restructuring**: Numerous import paths changed (e.g., `vllm.utils` → `vllm.utils.torch_utils` / `vllm.utils.math_utils`)
- **Attention Backend API Redesign**: New abstractions like `AttentionBackendEnum`, `AttentionSelectorConfig`, and `get_cudagraph_support()` were introduced
- **FusedMoE OOT (Out-of-Tree) Registration**: A new `CustomOp.register_oot` mechanism replaces monkey patching for custom op implementations
- **Compilation System Upgrade**: New components such as `TorchCompileWithNoGuardsWrapper` were added

To ensure the vLLM-Kunlun plugin works correctly on v0.15.1, we completed a comprehensive adaptation through **4 incremental PRs** (the 1/N series), covering the full pipeline from initialization fixes, API alignment, and code architecture refactoring to MoE kernel registration modernization.

---

## 2. Upgrade Overview

| # | PR | Theme | Files Changed | Lines Changed |
|:---:|:---|:---|:---:|:---:|
| 1/N | [#201](https://github.com/baidu/vLLM-Kunlun/pull/201) | Fix circular import causing plugin initialization failure | 4 files | +244 / -173 |
| 2/N | [#202](https://github.com/baidu/vLLM-Kunlun/pull/202) | Remove V0/V1 branching, fix import paths & dependency names | 6 files | +571 / -254 |
| 3/N | [#203](https://github.com/baidu/vLLM-Kunlun/pull/203) | Unify custom operator registration into `torch.ops` | 4 files | +2663 / -2634 |
| 4/N | [v0.15.1-4/N](https://github.com/xyDong0223/vLLM-Kunlun-kunlunops/commit/ede1a7fdd533f609207506c3341383dd67307c67) | Rewrite FusedMoE kernel registration using OOT method | 3 files | Refactor |
| **Total** | | | **17 files** | **~3500+ / -3000+** |

---

## 3. Detailed Phase Descriptions

### Phase 1 (PR #201): Fix Circular Imports — Make the Plugin Bootable

**Problem**: After upgrading to v0.15.1, the Kunlun plugin crashed during `register()` due to a circular import chain:

```
vllm_kunlun.__init__.register()
  → vllm_utils_wrapper.py
    → vllm.utils (weak_ref_tensor)
      → vLLM internal platform plugin loading
        → re-triggers vllm_kunlun.__init__ → Circular reference!
```

**Solution**:
- **Created `vllm_kunlun/schema.py`** (+98 lines): Extracted `direct_register_custom_op`, `patch_annotations_for_schema`, and related utilities from `vllm_utils_wrapper.py` into a standalone, zero-dependency module that can be safely imported during early plugin registration
- **Restructured `register()` entry point**: Removed all eager top-level imports that triggered circular dependencies; added structured `try/except` blocks with `logging` for better debuggability
- **Fixed import paths**: `from vllm.utils import weak_ref_tensor` → `from vllm.utils.torch_utils import weak_ref_tensor`

### Phase 2 (PR #202): Full API Alignment — Embrace V1-Only Architecture

**Key Changes**:
- **Removed all `VLLM_USE_V1` conditionals**: vLLM v0.15.1 is V1-only, so worker class selection and attention backend selection no longer require V0/V1 branching
- **Enhanced `patch_annotations_for_schema`**: Added support for PEP 604 union syntax (`T | None` → `Union[T, None]`) and return annotation patching to prevent `torch.library.infer_schema` failures
- **Updated `get_attn_backend_cls()` signature**: Migrated from positional parameters to the new `AttentionBackendEnum` + `AttentionSelectorConfig` API
- **Updated attention backend import paths**: `vllm.attention.backends.abstract` → `vllm.v1.attention.backend`; backend name changed from `"Kunlun_v1"` to `"CUSTOM"`
- **Added `TorchCompileWithNoGuardsWrapper`** (+279 lines): Guard-free `torch.compile` execution wrapper for v0.15.1 compilation
- **Deregistered models now supported upstream**: Qwen3 series models are natively supported in v0.15.1, so custom registrations were removed to avoid conflicts
- **Temporarily disabled quantization modules and MLA ops** (tracked as TODOs for v0.16.0)

### Phase 3 (PR #203): Code Architecture Refactor — Isolate Operator Registration

**Motivation**: Previously, 2600+ lines of custom operator registrations (`@custom_op` / `@impl` / `register_fake`) were appended to the end of `vllm_utils_wrapper.py`, a file whose primary responsibility was monkey-patching vLLM utility functions. This mixing of concerns made maintenance extremely difficult.

**Solution**:
- **Created `vllm_kunlun/ops/_custom_ops.py`** (+2644 lines): All custom operator registrations consolidated into a single, dedicated module covering 10+ categories and 50+ operators:

  | Category | Operators |
  |:---|:---|
  | Normalization | `rms_norm`, `fused_add_rms_norm`, etc. |
  | FP8 Quantization | `static_scaled_fp8_quant`, `dynamic_scaled_fp8_quant`, etc. |
  | INT8/INT4 Quantization | `scaled_int8_quant`, `quant2d`, `dequant_int4` |
  | Activation | `silu_and_mul`, `swigluoai_and_mul`, etc. |
  | MoE | `topk_softmax`, `moe_ffn_block`, etc. (10 operators) |
  | Attention | `I8_mqa_logits`, `fwd_kvcache_mla`, etc. |
  | GEMM | `matmul`, `cutlass_scaled_mm`, etc. |
  | LoRA | `sgmv_shrink_lora`, `bgmv_expand_lora`, etc. (7 operators) |
  | Sampling | `apply_repetition_penalties_`, `fast_topkv2` |

- **Cleaned up `vllm_utils_wrapper.py`** (-2633 lines): All operator registration code removed; the file now solely handles utility function monkey-patching
- **Further attention backend alignment**: Added `get_cudagraph_support()`, `supports_update_block_table`, and `dcp_world_size` parameter

### Phase 4 (v0.15.1-4/N): FusedMoE OOT Registration — Goodbye Monkey Patch

**Motivation**: Previously, Kunlun's FusedMoE adaptation was done via monkey patching — directly overwriting `vllm.model_executor.layers.fused_moe.layer.UnquantizedFusedMoEMethod` and `FusedMoE` at module level. This approach was fragile and unmaintainable. vLLM v0.15.1 introduced `CustomOp.register_oot`, allowing hardware vendors to register custom implementations through standard entry points.

**Key Changes**:
- **Adopted `@CustomOp.register_oot(name="UnquantizedFusedMoEMethod")` decorator** to replace monkey patching, leveraging vLLM's official OOT extension mechanism
- **Switched to Monolithic mode** (`is_monolithic = True`): `FusedMoE.forward_impl` calls `apply_monolithic(layer, x, router_logits)` directly, passing `router_logits` straight to KunlunOps for device-optimized routing on the chip
- **Deleted `KunlunFusedMoE` class** (-100+ lines): No longer necessary to subclass and override the entire `FusedMoE` class
- **Registration via `setup.py` entry points**: Added `vllm.plugins` entry point for plugin-based loading
- **Skipped Triton kernel initialization**: Overrode `process_weights_after_loading` to call `FusedMoEMethodBase`'s base method directly, bypassing Triton kernel setup

---

## 4. Architecture Comparison

```
                    Before (v0.14.x)                          After (v0.15.1)
             ┌──────────────────────────┐            ┌──────────────────────────┐
  Init       │ register() eagerly       │            │ register() lightweight   │
             │ imports wrapper          │   ──→      │ imports schema.py        │
             │ → circular import ❌      │            │ → safe init ✅            │
             ├──────────────────────────┤            ├──────────────────────────┤
  V0/V1      │ if VLLM_USE_V1:          │            │ V1-only path             │
  Branching  │   use V1 worker          │   ──→      │ No conditional branches  │
             │ else:                    │            │ Cleaner code ✅           │
             │   use V0 worker          │            │                          │
             ├──────────────────────────┤            ├──────────────────────────┤
  Custom Op  │ All registrations in     │            │ Dedicated module         │
  Registry   │ vllm_utils_wrapper.py    │   ──→      │ ops/_custom_ops.py       │
             │ (2700+ line mixed file)  │            │ (single responsibility ✅)│
             ├──────────────────────────┤            ├──────────────────────────┤
  FusedMoE   │ Monkey Patch overriding  │            │ @CustomOp.register_oot   │
             │ layer.FusedMoE           │   ──→      │ + is_monolithic          │
             │ layer.UnquantizedMethod  │            │ + entry_points ✅         │
             └──────────────────────────┘            └──────────────────────────┘
```

---

## 5. Known TODOs & Future Plans

| Item | Status | Owner | Target Version |
|:---|:---:|:---|:---:|
| Re-enable quantization modules (AWQ / GPTQ / INT8, etc.) | 🔴 Disabled | @liwei109 | v0.16.0 |
| Re-enable MLA operators | 🔴 Disabled | @xyDong0223 | v0.16.0 |
| GLM5 model config patch | 🟡 Disabled | @xyDong0223 | TBD |
| `ModelConfig.is_deepseek_mla` patch | 🟡 Disabled | @xyDong0223 | TBD |
| `use_cudagraph` custom_ops override | 🟡 Disabled | — | Pending validation |

---

## 6. Conclusion

This v0.15.1 upgrade represents the largest framework adaptation effort in vLLM-Kunlun's history. Through 4 incremental PRs, we achieved:

1. **🔧 Fixed a Critical Blocker**: Resolved the circular import issue that completely prevented plugin initialization
2. **🏗️ Completed Architectural Refactoring**: Clean separation of concerns across operator registration, utility functions, and FusedMoE layer
3. **🚀 Embraced Upstream Best Practices**: Migrated from monkey patching to OOT registration; simplified from V0/V1 dual-path to V1-only
4. **📐 Full v0.15.1 API Alignment**: Import paths, function signatures, backend naming, and compilation system are all adapted
5. **🔮 Laid the Foundation for Future Versions**: Clear module boundaries will make re-enabling quantization ops and MLA in v0.16.0 significantly smoother

**Comment by liyandong001** [comment]
Could you please clarify whether vLLM-Kunlun plans to support community-developed custom Kunlun XPU operators? Currently, it looks like core ops are provided through prebuilt `kunlun_ops` wheels, so external developers cannot easily implement, compile, and integrate their own kernels.
